crypto/src/utils.rs

```diff
@@ -60,63 +61,78 @@ pub fn read_keys_from_file<P: AsRef<Path>>(
     key_file.read_to_end(&mut file_content)?;
     let keys: EncryptedKeys =
         toml::from_slice(file_content.as_slice()).map_err(|e| Error::new(ErrorKind::Other, e))?;
-    let pub_key = PublicKey::from_slice(
-        &hex::decode(&keys.public_key).map_err(|e| Error::new(ErrorKind::Other, e))?,
-    ).ok_or_else(|| Error::new(
-        ErrorKind::Other,
-        "Couldn't create PublicKey from slice",
-    ))?;
-    let sec_key = SecretKey::from_slice(&decrypt(&keys, pass_phrase)?).ok_or_else(|| Error::new(
-        ErrorKind::Other,
-        "Couldn't create SecretKey from slice",
-    ))?;
+    let pub_key = keys.public_key;
+    let sec_key = keys.decrypt(pass_phrase)?;
 
     Ok((pub_key, sec_key))
 }
 
+struct PublicKeyHex;
+
+impl Hex<PublicKey> for PublicKeyHex {
+    fn create_bytes(value: &PublicKey) -> Cow<[u8]> {
+        Cow::Borrowed(&*value.as_ref())
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Result<PublicKey, String> {
+        PublicKey::from_slice(bytes)
+            .ok_or_else(|| "Couldn't create PublicKey from slice".to_owned())
+    }
+}
+
 #[derive(Serialize, Deserialize)]
 struct EncryptedKeys {
-    public_key: String,
+    #[serde(with = "PublicKeyHex")]
+    public_key: PublicKey,
     secret_key: ErasedPwBox,
 }
 
-fn encrypt(
-    public_key: String,
-    secret_key: &[u8],
-    pass_phrase: &[u8],
-) -> Result<EncryptedKeys, Error> {
-    let mut rng = thread_rng();
-    let mut eraser = Eraser::new();
-    eraser.add_suite::<Sodium>();
-    let pwbox = Sodium::build_box(&mut rng)
-        .seal(pass_phrase, secret_key)
-        .map_err(|_| Error::new(ErrorKind::Other, "Couldn't create a pw box"))?;
-    let encrypted_key = eraser
-        .erase(pwbox)
-        .map_err(|_| Error::new(ErrorKind::Other, "Couldn't convert a pw box"))?;
-
-    Ok(EncryptedKeys {
-        public_key,
-        secret_key: encrypted_key,
-    })
-}
+impl EncryptedKeys {
+    fn encrypt(
+        public_key: PublicKey,
```

Comment:
I don't think that we need to transfer ownership here, public key should be passed by reference.

For example, the user might want to encrypt the same pair of keys with several passwords. Yeah, I know it's weird and nobody does that. But we ought not to make a new API inconsistent right at the beginning.

Reply:
FYI `PublicKey` derives Copy and Clone traits.